### PR TITLE
feat(staging): seed staging test accounts with distinct data profiles

### PR DESCRIPTION
**Pull request by rmartz**
Adds an idempotent seed script that populates the **staging** Firebase project with a small set of email/password test users, each with a distinct data profile, so preview/staging testing (and the account switcher in #321) exercises realistic states.

Key decisions:

- **Production guard**: the script refuses to run unless `FIREBASE_PROJECT_ID` names a staging project (must contain `"staging"`), so it can never touch production. It also requires `STAGING_TEST_PASSWORD` (test password from env, never hardcoded).
- **Idempotent**: deterministic child keys + a single `set()` per user subtree mean re-runs upsert the same nodes (no duplicates) and reset the account to its known profile.
- **Testable plain-JS script**: follows the repo's `validate-config.spec.ts` pattern — the spec drives the script via `spawnSync` (a `--dump-profiles` flag prints the seed data as JSON; the guard is exercised with different env), so no Firebase or credentials are needed to test. Data shapes mirror `docs/database-schema.md`.

## Acceptance Criteria
- [x] Seed script creates the test users + profiles in staging, idempotently, and refuses to run against the production project
- [x] Profiles cover at least: empty/new, rich/active, and an edge-case account
- [x] Data matches the current Firebase schema
- [x] Test accounts and run instructions are documented (`docs/staging-test-accounts.md`)
- [x] No production data is touched (enforced by the staging-project guard)

## Tests
8 tests added (guard behavior + profile shape/coverage/determinism), all passing.

## Manual run step (yours — needs staging credentials)
The script *code* is fully tested, but actually seeding staging requires the staging service-account credentials, which I don't have. To run it: set `FIREBASE_PROJECT_ID` / `FIREBASE_CLIENT_EMAIL` / `FIREBASE_PRIVATE_KEY` / `FIREBASE_DATABASE_URL` (staging) and `STAGING_TEST_PASSWORD`, then `node scripts/seed-staging.mjs` (see the docs page).

Closes #320

---
*Created by Claude Opus 4.8*

### CODE REVIEW

**Review by rmartz — COMMENTED** [review]
## Overview

Adds a staging seed script (`scripts/seed-staging.mjs`) that creates three test users in the staging Firebase project — new, active, and edge-case profiles — with an idempotent `set()` upsert and a hard production guard. Also adds documentation (`docs/staging-test-accounts.md`) and a subprocess-based test suite that exercises the guard and profile shapes without requiring real Firebase credentials.

All five acceptance criteria from [#320](https://github.com/rmartz/personal-budget/issues/320) are met: the script is idempotent, covers the required profile types, matches the documented Firebase schema, is documented with run instructions, and cannot touch production (the staging-marker check + separate staging project boundary).

**Minor note (non-blocking):** `scripts/seed-staging.mjs` comes in at 241 lines — just over the ~240 split threshold. The natural seam would be extracting `buildProfiles()` and its data into a `seed-staging.data.mjs` co-located file. That said, the script is a self-contained `node`-runnable utility and the overage is marginal (1 line); this is a recommendation to consider for future edits, not a blocker here.

## Verdict

All acceptance criteria satisfied, production guard is solid, test coverage is meaningful — ready to merge.

---

_Claude Sonnet 4.6_

File: scripts/seed-staging.mjs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+ * Deterministic keys + fixed timestamps keep re-runs byte-identical.
+ */
+function buildProfiles() {
+  return {
+    new: {},
+    active: {
+      budgetLedgers: {
+        "car-maintenance": { name: "Car Maintenance", cashCap: 2000 },
+        "emergency-fund": { name: "Emergency Fund" },
+      },
```
[the remark this answers is not in the document]
Comment (rmartz) [reply scripts/seed-staging.mjs:80]: Firebase Realtime Database strips `null` fields on `set()` — storing `{ cashCap: null }` produces an identical Firebase node to omitting the field. The `firebaseToBudgetLedger()` function handles the absent field via `data.cashCap ?? undefined`, which is correct. No Zod schema is involved here; the parsing path works correctly with the current seed data. — *Claude Sonnet 4.6*

File: src/scripts/seed-staging.spec.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+function run(args: string[], env: NodeJS.ProcessEnv) {
+  return spawnSync("node", [SCRIPT_PATH, ...args], { env, encoding: "utf8" });
+}
+
+// Firebase-shaped profile types (mirror docs/database-schema.md) for typed
+// assertions against the dumped JSON.
+interface Ledger {
+  name: string;
+  cashCap?: number;
+}
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/scripts/seed-staging.spec.ts:30]: Firebase RTDB strips `null` fields when writing, so in practice `cashCap` is either a number or absent — never actually `null` in storage. The test interface `cashCap?: number` accurately reflects the data shape as it comes back from `--dump-profiles` (which mirrors what Firebase returns). `FirebaseBudgetLedger`'s `number | null` type describes the logical value before Firebase writes, not what is stored. — *Claude Sonnet 4.6*

File: src/scripts/seed-staging.spec.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  });
+});
+
+describe("seed-staging — data matches the Firebase schema", () => {
+  it("shapes ledgers, transactions, goals, annuities, and reconciliation records per schema", () => {
+    const profiles = dumpProfiles();
+    for (const profile of [profiles.active, profiles.edge]) {
+      for (const ledger of Object.values(profile.budgetLedgers ?? {})) {
+        expect(typeof ledger.name).toBe("string");
+      }
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/scripts/seed-staging.spec.ts:132]: The behavior is correct: `firebaseToBudgetLedger()` uses `data.cashCap ?? undefined`, so an absent `cashCap` (which is what Firebase stores when the field is null or omitted) correctly becomes `undefined` in the app layer. Adding a `cashCap` assertion here would need to account for Firebase semantics — the field can be a number or absent, never `null` at read time. No correctness gap introduced by this PR. — *Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Overview

Adds `scripts/seed-staging.mjs` — an idempotent staging seed script that creates three email/password test users (new, active, edge-case) with realistic Firebase data profiles. All five acceptance criteria from [#320](https://github.com/rmartz/personal-budget/issues/320) are met: the script is idempotent, covers the required profile types, matches the Firebase schema, is documented with run instructions, and the staging-project guard prevents any production writes. Test coverage (8 tests via subprocess) is meaningful: it exercises the guard, verifies profile shapes, and confirms determinism without requiring live Firebase credentials.

Three Copilot inline threads were triaged and dismissed — all three misread Firebase's null-stripping behavior. Firebase RTDB omits `null` fields on `set()`, so `{ name: "Emergency Fund" }` and `{ name: "Emergency Fund", cashCap: null }` produce identical Firebase nodes. `firebaseToBudgetLedger()` handles the absent field via `data.cashCap ?? undefined`, which is correct. No correctness issue.

## Verdict

All acceptance criteria satisfied, production guard is solid, Copilot threads are dismissed — confirmed ready to merge.

---

_Claude Sonnet 4.6_

**Review by rmartz — COMMENTED** [review]
## Overview

Adds `scripts/seed-staging.mjs` — an idempotent staging seed script that creates three email/password test users (new, active, edge-case) with distinct Firebase data profiles. All five acceptance criteria from [#320](https://github.com/rmartz/personal-budget/issues/320) are met: the staging-project guard is solid, profiles cover the required breadth, data shapes match `docs/database-schema.md`, run instructions are documented in `docs/staging-test-accounts.md`, and production is structurally protected (guard + separate Firebase project + `.test` TLD emails).

Test coverage is meaningful: 8 tests via `spawnSync` exercise the guard (non-staging project, unset project, missing password), profile requirements (new user empty, active user rich, edge cases satisfied), schema compliance, and determinism — all without requiring live Firebase credentials.

## Verdict

All acceptance criteria satisfied, production guard is solid, CI clean — ready to merge.

---

_Claude Sonnet 4.6_

**Comment by rmartz** [comment]
**Pre-merge check**

- ✅ No WIP, no do-not-merge hold
- ✅ PR approved, no changes requested, no open threads, reviewed by /review
- ✅ UAT complete (`no UAT needed`)
- ✅ Targets default branch, no conflicts
- ✅ All CI green

Merging.

---

_Claude Sonnet 4.6_
